src/math/exp_fcn_hifi4.c

```diff
@@ -0,0 +1,222 @@
+ * Return Type	: void
+ * Description:Perform element-wise multiplication on in_0 and in_1
+ * while keeping the required product word length and fractional
+ * length in mind. mul_s64 function divide the 64-bit quantities
+ * into two 32-bit words,multiply the low words to produce the
+ * lowest and second-lowest words in the result, then both pairs
+ * of low and high words from different numbers to produce the
+ * second and third lowest words in the result, and finally both
+ * high words to produce the two highest words in the outcome.
+ * Add them all up, taking carry into consideration.
+ *
+ * The 64 x 64 bit multiplication of operands in_0 and in_1 is
+ * shown in the image below. The 64-bit operand in_0,in_1 is
+ * represented by the notation in0_H, in1_H for the top 32 bits
+ * and in0_L, in1_L for the bottom 32 bits.
+ *
+ *				in0_H : in0_L
+ *			x	in1_H : in1_L
+ *			---------------------
+ *	P0			in0_L x in1_L
+ *	P1		in0_H x in1_L		64 bit inner multiplication
+ *	P2		in0_L x in1_H		64 bit inner multiplication
+ *	P3	in0_H x in1_H
+ *			--------------------
+ *			[64 x 64 bit multiplication] sum of inner products
+ * All combinations are multiplied by one another and then added.
+ * Each inner product is moved into its proper power location.given the names
+ * of the inner products, redoing the addition where 000 represents 32 zero
+ * bits.The inner products can be added together in 64 bit addition.The sum
+ * of two 64-bit numbers yields a 65-bit output.
+ *		   (P0H:P0L)
+ *		P1H(P1L:000)
+ *		P2H(P2L:000)
+ *	P3H:P3L(000:000)
+ *	.......(aaa:P0L)
+ * By combining P0H:P0L and P1L:000. This can lead to a carry, denote as CRY0.
+ * The partial result is then multiplied by P2L:000.
+ * We call it CRY1 because it has the potential to carry again.
+ *	(CRY0 + CRY1)P0H:P0L
+ *	(	 P1H)P1L:000
+ *	(	 P2H)P2L:000
+ *	(P3H:	 P3L)000:000
+ *	--------------------
+ *	(ccc:bbb)aaa:P0L
+ * P1H, P2H, and P3H:P3L are added to the carry CRY0 + CRY1.This increase will
+ * not result in an overflow.
+ *
+ */
+static void mul_s64(ae_int64 in_0, ae_int64 in_1, ae_int64 *ptroutbitshi, ae_int64 *ptroutbitslo)
+{
+	ae_int64 producthihi, producthilo, productlolo;
+	ae_int64 producthi, carry, product_hl_lh_h, product_hl_lh_l;
+
+	ae_int32x2 in0_32 = AE_MOVINT32X2_FROMINT64(in_0);
+	ae_int32x2 in1_32 = AE_MOVINT32X2_FROMINT64(in_1);
+
+	ae_ep ep_lolo = AE_MOVEA(0);
+	ae_ep ep_hilo = AE_MOVEA(0);
+	ae_ep ep_HL_LH = AE_MOVEA(0);
+
+	producthihi = AE_MUL32_HH(in0_32, in1_32);
+	/* Input-32x32-bit(in1_32xin0_32)into 72-bit multiplication operations */
+	/* Output-lower 64 bits of the result are stored in producthilo */
+	/* Output-upper eight bits are stored in ep_hilo */
+	AE_MULZAAD32USEP_HL_LH(ep_hilo, producthilo, in1_32, in0_32);
+	productlolo = AE_MUL32U_LL(in0_32, in1_32);
+
+	product_hl_lh_h = AE_SRAI72(ep_hilo, producthilo, 32);
+	product_hl_lh_l = AE_SLAI64(producthilo, 32);
+
+	AE_ADD72(ep_lolo, productlolo, ep_HL_LH, product_hl_lh_l);
```

Comment:
@ShriramShastry can we do the same here that we did for `AE_MULZAAD32USEP_HL_LH`? Unfortunetly this is hifi being bad that requires this

Reply:
`AE_MULZAAD32USEP_HL_LH` is an intrinsic multiply add, whereas AE_ADD72 is a bit increment. 

The `AE_MULZAAD32USEP_HL_LH` is a `32x32`-bit unsigned times signed integer dual MAC with a result of `72` bits and `no saturation`.

`AE_ADD72` is a 72-bit increment of an `AE_EP`, `AE_DR` register pair by another 72-bit register pair.

Could you perhaps provide a numerical example to back up your abd opinion `AE_ADD72`?

Thank you

Reply:
@cujomalainey , any feedback here ? Thanks

Reply:
@ShriramShastry im just asking for comment (in the code) explaining whats going on with the variables since HiFi violates C standard here so its hard to tell what is the input and what is the output

Reply:
@ShriramShastry I don't see the comment for `AE_ADD72`, did you push?